shell/browser/extensions/api/tabs/tabs_api.cc

```diff
@@ -319,4 +322,182 @@ ExtensionFunction::ResponseAction TabsSetZoomSettingsFunction::Run() {
   return RespondNow(NoArguments());
 }
 
+bool IsKillURL(const GURL& url) {
+#if DCHECK_IS_ON()
+  // Caller should ensure that |url| is already "fixed up" by
+  // url_formatter::FixupURL, which (among many other things) takes care
+  // of rewriting about:kill into chrome://kill/.
+  if (url.SchemeIs(url::kAboutScheme))
+    DCHECK(url.IsAboutBlank() || url.IsAboutSrcdoc());
+#endif
+
+  static const char* const kill_hosts[] = {
+      chrome::kChromeUICrashHost,         chrome::kChromeUIDelayedHangUIHost,
+      chrome::kChromeUIHangUIHost,        chrome::kChromeUIKillHost,
+      chrome::kChromeUIQuitHost,          chrome::kChromeUIRestartHost,
+      content::kChromeUIBrowserCrashHost, content::kChromeUIMemoryExhaustHost,
+  };
+
+  if (!url.SchemeIs(content::kChromeUIScheme))
+    return false;
+
+  return base::Contains(kill_hosts, url.host_piece());
+}
+
+GURL ResolvePossiblyRelativeURL(const std::string& url_string,
+                                const Extension* extension) {
+  GURL url = GURL(url_string);
+  if (!url.is_valid() && extension)
+    url = extension->GetResourceURL(url_string);
+
+  return url;
+}
+bool PrepareURLForNavigation(const std::string& url_string,
+                             const Extension* extension,
+                             GURL* return_url,
+                             std::string* error) {
+  GURL url = ResolvePossiblyRelativeURL(url_string, extension);
+
+  // Ideally, the URL would only be "fixed" for user input (e.g. for URLs
+  // entered into the Omnibox), but some extensions rely on the legacy behavior
+  // where all navigations were subject to the "fixing".  See also
+  // https://crbug.com/1145381.
+  url = url_formatter::FixupURL(url.spec(), "" /* = desired_tld */);
+
+  // Reject invalid URLs.
+  if (!url.is_valid()) {
+    const char kInvalidUrlError[] = "Invalid url: \"*\".";
+    *error = ErrorUtils::FormatErrorMessage(kInvalidUrlError, url_string);
+    return false;
+  }
+
+  // Don't let the extension crash the browser or renderers.
+  if (IsKillURL(url)) {
+    const char kNoCrashBrowserError[] =
+        "I'm sorry. I'm afraid I can't do that.";
```

Comment:
😂 I actually like this message, but would it be possible for us to give the user more of a hint about what went wrong? Looking at `kill_hosts[]` above, would a user be trying to intentionally kill their tab/app by passing one of these URLs in?

Reply:
This is actually copy-pasted directly from Chromium: https://source.chromium.org/chromium/chromium/src/+/main:chromecast/browser/extensions/api/tabs/tabs_constants.cc;l=84;drc=479cc17585a64910853e9949b053499ecbeca9a5

This only happens if an extension tries to do something like `chrome.tabs.update(tabId, { url: 'chrome://crash' })`.